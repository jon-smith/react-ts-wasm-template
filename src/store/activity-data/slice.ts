import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { ActivityContainer, fromJSONData } from 'library/activity-data/activity-container';
import ExampleData from 'library/activity-data/data/example-over-unders';
import { Mutable } from 'library/utils/type-utils';

type ExtendedActivityContainer = ActivityContainer & { filename: string };

export type View = 'data' | 'interval-detector';

export type ActivityState = Readonly<{
	activities: ExtendedActivityContainer[];
	selectedIndex?: number;
	view: View;
}>;

const defaultState: ActivityState = {
	activities: [],
	view: 'data',
};

function addActivitiesImpl(
	state: Mutable<ActivityState>,
	newActivities: ExtendedActivityContainer[]
) {
	const previousLength = state.activities.length;
	state.activities = [...state.activities, ...newActivities];
	// Select the first activity we have loaded
	if (state.activities.length > previousLength) {
		state.selectedIndex = previousLength;
	}
}

const slice = createSlice({
	name: 'activityData',
	initialState: defaultState,
	reducers: {
		loadExampleData(state) {
			addActivitiesImpl(state, [{ ...fromJSONData(ExampleData), filename: 'Example.json' }]);
		},
		setView(state, action: PayloadAction<View>) {
			state.view = action.payload;
		},
		addActivities(state, action: PayloadAction<ExtendedActivityContainer[]>) {
			addActivitiesImpl(state, action.payload);
		},
		clearActivityData(state) {
			state.activities = [];
		},
		setSelectedIndex(state, action: PayloadAction<number | undefined>) {
			state.selectedIndex = action.payload;
		},
	},
});

export const { reducer, actions } = slice;

export const {
	loadExampleData,
	setView,
	addActivities,
	clearActivityData,
	setSelectedIndex,
} = actions;
