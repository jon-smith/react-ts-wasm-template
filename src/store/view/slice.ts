import { createSlice, PayloadAction } from '@reduxjs/toolkit';

export type Page = 'data' | 'workout-creator';

export interface ViewState {
	readonly currentPage: Page;
}

const defaultState: ViewState = {
	currentPage: 'data',
};

const slice = createSlice({
	name: 'view',
	initialState: defaultState,
	reducers: {
		setCurrentPage(state, action: PayloadAction<Page>) {
			state.currentPage = action.payload;
		},
	},
});

export const { reducer, actions } = slice;

export const { setCurrentPage } = actions;
