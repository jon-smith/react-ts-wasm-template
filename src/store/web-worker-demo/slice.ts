import { createSlice, PayloadAction, createAsyncThunk } from '@reduxjs/toolkit';
import { AwaitedType } from 'library/utils/type-utils';
import { runWebWorker } from './worker-caller';

type WorkerResultT = AwaitedType<ReturnType<typeof runWebWorker>>;
type WorkerInput = string;

export type WebWorkerDemoState = Readonly<{
	input: WorkerInput;
	isGenerating: boolean;
	processedData: { series: WorkerResultT; input?: WorkerInput };
}>;

const defaultState: WebWorkerDemoState = {
	input: '',
	isGenerating: false,
	processedData: {
		series: '',
	},
};

export function dataProcessingRequired(state: WebWorkerDemoState) {
	if (state.processedData.input === undefined) return true;

	return state.processedData.input !== state.input;
}

export const processData = createAsyncThunk(
	'webWorkerDemo/runWorker',
	async (input: WorkerInput) => {
		const series = await runWebWorker(input);
		return {
			series,
			input,
		};
	}
);

const webWorkerDemoSlice = createSlice({
	name: 'webWorkerDemo',
	initialState: defaultState,
	reducers: {
		setInput(state, action: PayloadAction<WorkerInput>) {
			state.input = action.payload;
		},
	},
	extraReducers: (builder) => {
		builder.addCase(processData.pending, (state) => {
			state.isGenerating = true;
		});
		builder.addCase(processData.rejected, (state) => {
			state.isGenerating = false;
		});
		builder.addCase(processData.fulfilled, (state, { payload }) => {
			state.processedData = payload;
			state.isGenerating = false;
		});
	},
});

export const { reducer, actions } = webWorkerDemoSlice;

export const { setInput } = actions;
