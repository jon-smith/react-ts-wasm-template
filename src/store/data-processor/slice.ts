import { createSlice, PayloadAction, createAsyncThunk } from '@reduxjs/toolkit';
import { getSmoothedTimeSeries } from 'library/activity-data/activity-calculator';
import { ActivityContainer } from 'library/activity-data/activity-container';
import { clamp } from 'library/utils/math-utils';
import { performDataSmoothing } from './worker-caller';

type TimeSeriesT = ReturnType<typeof getSmoothedTimeSeries>;

type DataProcessingInput = { smoothingRadius: number; activity?: ActivityContainer };

export type DataProcessorState = Readonly<{
	smoothingRadius: number;
	isGenerating: boolean;
	processedData: { series: TimeSeriesT; input?: DataProcessingInput };
}>;

const defaultState: DataProcessorState = {
	smoothingRadius: 0,
	isGenerating: false,
	processedData: {
		series: [],
	},
};

export function dataSmoothingRequired(
	state: DataProcessorState,
	activity: ActivityContainer | undefined
) {
	if (state.processedData.input === undefined) return true;

	return (
		state.processedData.input.activity !== activity ||
		state.processedData.input.smoothingRadius !== state.smoothingRadius
	);
}

export const smoothData = createAsyncThunk(
	'dataProcessing/smoothData',
	async (input: DataProcessingInput) => {
		const series = await performDataSmoothing(input.activity, input.smoothingRadius);
		return {
			series,
			input,
		};
	}
);

const dataProcessorSlice = createSlice({
	name: 'dataProcessing',
	initialState: defaultState,
	reducers: {
		setSmoothingRadius(state, action: PayloadAction<number>) {
			state.smoothingRadius = clamp(action.payload, 1, 600);
		},
	},
	extraReducers: (builder) => {
		builder.addCase(smoothData.pending, (state) => {
			state.isGenerating = true;
		});
		builder.addCase(smoothData.rejected, (state) => {
			state.isGenerating = false;
		});
		builder.addCase(smoothData.fulfilled, (state, { payload }) => {
			state.processedData = payload;
			state.isGenerating = false;
		});
	},
});

export const { reducer, actions } = dataProcessorSlice;

export const { setSmoothingRadius } = actions;
