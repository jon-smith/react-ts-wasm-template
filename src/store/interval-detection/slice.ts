import { createSlice, PayloadAction, createAsyncThunk } from '@reduxjs/toolkit';
import {
	IntervalDetectionParameters,
	IntervalDetectionResults,
} from 'library/activity-data/interval-detection';
import { ActivityContainer } from 'library/activity-data/activity-container';
import { clamp } from 'library/utils/math-utils';
import { performIntervalDetection } from './worker-caller';

type IntervalGenerationInput = {
	activity: ActivityContainer | undefined;
	params: IntervalDetectionParameters;
};

export type IntervalDetectionState = Readonly<{
	generationParams: IntervalDetectionParameters;
	ftp: number;
	isGenerating: boolean;
	detectionResults: { results: IntervalDetectionResults; input?: IntervalGenerationInput };
}>;

const defaultState: IntervalDetectionState = {
	ftp: 200,
	generationParams: {
		minIntervalDuration: 10,
		stepThreshold: 10,
		windowRadius: 10,
		discrepencySmoothingRadius: 1,
		inputSmoothingRadius: 1,
	},
	isGenerating: false,
	detectionResults: {
		results: {
			intervals: [],
			rawInput: [],
			smoothedInput: [],
			discrepencyCurve: [],
			detectedStepTimePoints: [],
		},
	},
};

export function generateIntervalsRequired(
	state: IntervalDetectionState,
	activity: ActivityContainer | undefined
) {
	if (state.detectionResults.input === undefined) return true;

	return (
		state.detectionResults.input.activity !== activity ||
		state.detectionResults.input.params !== state.generationParams
	);
}

export const generateIntervals = createAsyncThunk(
	'intervalDetection/generateIntervals',
	// Note this function doesn't actually run asynchronously at the moment
	// but I intend to use a worker thread in the future
	// For now I just wanted to try out the usage of createAsyncThunk
	async (input: IntervalGenerationInput) => {
		const results = await performIntervalDetection(input.activity, input.params);
		return {
			results,
			input,
		};
	}
);

const intervalDetectionSlice = createSlice({
	name: 'intervalDetection',
	initialState: defaultState,
	reducers: {
		setFTP(state, action: PayloadAction<number>) {
			state.ftp = clamp(action.payload, 1, 999);
		},
		setWindowRadius(state, action: PayloadAction<number>) {
			state.generationParams.windowRadius = clamp(action.payload, 1, 600);
		},
		setStepThreshold(state, action: PayloadAction<number>) {
			state.generationParams.stepThreshold = clamp(action.payload, 0, 30);
		},
		setInputSmoothing(state, action: PayloadAction<number>) {
			state.generationParams.inputSmoothingRadius = clamp(action.payload, 0, 60);
		},
		setDiscrepencySmoothing(state, action: PayloadAction<number>) {
			state.generationParams.discrepencySmoothingRadius = clamp(action.payload, 0, 60);
		},
	},
	extraReducers: (builder) => {
		builder.addCase(generateIntervals.pending, (state) => {
			state.isGenerating = true;
		});
		builder.addCase(generateIntervals.rejected, (state) => {
			state.isGenerating = false;
		});
		builder.addCase(generateIntervals.fulfilled, (state, { payload }) => {
			state.detectionResults = payload;
			state.isGenerating = false;
		});
	},
});

export const { reducer, actions } = intervalDetectionSlice;

export const {
	setFTP,
	setWindowRadius,
	setStepThreshold,
	setInputSmoothing,
	setDiscrepencySmoothing,
} = actions;
