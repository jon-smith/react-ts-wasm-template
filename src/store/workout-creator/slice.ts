import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import * as ArrayUtils from 'library/utils/array-utils';
import { Mutable } from 'library/utils/type-utils';
import { Interval } from 'library/activity-data/interval';
import { WorkoutCreatorState } from './types';

const defaultIntervals: Interval[] = [
	{ intensityPercent: 0.3, durationSeconds: 60 },
	{ intensityPercent: 0.4, durationSeconds: 60 },
	{ intensityPercent: 0.5, durationSeconds: 60 },
	{ intensityPercent: 0.6, durationSeconds: 60 * 5 },
	...Array(13)
		.fill([
			{ intensityPercent: 1.3, durationSeconds: 30 },
			{ intensityPercent: 0.6, durationSeconds: 15 },
		])
		.flat(),
	{ intensityPercent: 0.6, durationSeconds: 60 * 5 },
];

const defaultState: WorkoutCreatorState = {
	ftp: 200,
	newInterval: { intensityPercent: 1.0, durationSeconds: 0 },
	currentIntervals: defaultIntervals,
	history: [defaultIntervals],
	currentHistoryPosition: 0,
	selectedIndex: null,
};

function setIntervalsImpl(state: Mutable<WorkoutCreatorState>, intervals: Interval[]) {
	const areEqual = (a: Interval, b: Interval) =>
		a.intensityPercent === b.intensityPercent && a.durationSeconds === b.durationSeconds;
	if (!ArrayUtils.areEqual(intervals, state.currentIntervals, areEqual)) {
		const newHistory = [...state.history.slice(0, state.currentHistoryPosition + 1), intervals];

		state.history = newHistory;
		state.currentHistoryPosition = newHistory.length - 1;
		state.currentIntervals = intervals;
	}
}

const workoutCreatorSlice = createSlice({
	name: 'workoutCreator',
	initialState: defaultState,
	reducers: {
		setIntervals(state, action: PayloadAction<Interval[]>) {
			setIntervalsImpl(state, action.payload);
		},
		undo(state) {
			if (state.currentHistoryPosition > 0) {
				state.currentIntervals = state.history[state.currentHistoryPosition - 1];
				state.currentHistoryPosition -= 1;
			}
		},
		redo(state) {
			if (state.currentHistoryPosition < state.history.length - 1) {
				state.currentIntervals = state.history[state.currentHistoryPosition + 1];
				state.currentHistoryPosition += 1;
			}
		},
		setSelectedIndex(state, action: PayloadAction<number | null>) {
			state.selectedIndex = action.payload;
		},
		setSelectedIntensity(state, action: PayloadAction<number>) {
			if (state.selectedIndex === null) {
				state.newInterval.intensityPercent = action.payload;
			} else {
				const updatedIntervals = state.currentIntervals.slice();
				updatedIntervals[state.selectedIndex] = {
					...updatedIntervals[state.selectedIndex],
					intensityPercent: action.payload,
				};
				setIntervalsImpl(state, updatedIntervals);
			}
		},
		setSelectedLength(state, action: PayloadAction<number>) {
			if (state.selectedIndex === null) {
				state.newInterval.durationSeconds = action.payload;
			} else {
				const updatedIntervals = state.currentIntervals.slice();
				updatedIntervals[state.selectedIndex] = {
					...updatedIntervals[state.selectedIndex],
					durationSeconds: action.payload,
				};
				setIntervalsImpl(state, updatedIntervals);
			}
		},
		setFTP(state, action: PayloadAction<number>) {
			state.ftp = action.payload;
		},
	},
});

export const { reducer, actions } = workoutCreatorSlice;

export const {
	setIntervals,
	undo,
	redo,
	setSelectedIndex,
	setSelectedIntensity,
	setSelectedLength,
	setFTP,
} = actions;
