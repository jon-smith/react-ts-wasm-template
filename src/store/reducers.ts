import { combineReducers } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import { ActivityState, reducer as activityReducer } from './activity-data/slice';
import { ViewState, reducer as viewReducer } from './view/slice';
import { WorkoutCreatorState } from './workout-creator/types';
import { reducer as workoutCreatorReducer } from './workout-creator/slice';
import {
	IntervalDetectionState,
	reducer as intervalDetectionReducer,
} from './interval-detection/slice';

export const rootReducer = combineReducers({
	activities: activityReducer,
	intervalDetection: intervalDetectionReducer,
	view: viewReducer,
	workoutCreator: workoutCreatorReducer,
});

export type RootState = ReturnType<typeof rootReducer>;

export const useRootSelector = <T extends unknown>(selector: (s: RootState) => T) =>
	useSelector<RootState, T>((s) => selector(s));

export const useActivitySelector = <T extends unknown>(selector: (s: ActivityState) => T) =>
	useSelector<RootState, T>((s) => selector(s.activities));

export const useIntervalDetectionSelector = <T extends unknown>(
	selector: (s: IntervalDetectionState) => T
) => useSelector<RootState, T>((s) => selector(s.intervalDetection));

export const useViewSelector = <T extends unknown>(selector: (s: ViewState) => T) =>
	useSelector<RootState, T>((s) => selector(s.view));

export const useWorkoutCreatorSelector = <T extends unknown>(
	selector: (s: WorkoutCreatorState) => T
) => useSelector<RootState, T>((s) => selector(s.workoutCreator));
