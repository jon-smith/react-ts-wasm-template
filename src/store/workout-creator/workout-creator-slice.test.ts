import { Action } from 'redux';
import { rootReducer } from '.././reducers';
import { actions } from './slice';

// Use the worker-caller mock which doesn't use web-workers
jest.mock('.././interval-detection/worker-caller');

const getInitialState = () => rootReducer(undefined, {} as Action);

describe('Workout Creator Store', () => {
	it('initial state should match snapshot', () => {
		const initialState = getInitialState();
		expect(initialState.workoutCreator).toMatchSnapshot();
	});

	it('updates intervals', () => {
		const store = getInitialState();

		const intervalsToSet = [{ color: '', durationSeconds: 60, intensityPercent: 1.2 }];
		const updatedStore = rootReducer(store, actions.setIntervals(intervalsToSet));

		expect(updatedStore.workoutCreator.currentIntervals).toEqual(intervalsToSet);
		expect(updatedStore.workoutCreator.history.length).toEqual(2);
		expect(updatedStore.workoutCreator.currentHistoryPosition).toEqual(1);
	});

	it('doesnt change intervals if set to same', () => {
		const store = getInitialState();
		const intervalCopy = store.workoutCreator.currentIntervals.slice();

		const updatedStore = rootReducer(store, actions.setIntervals(intervalCopy));

		// Confirm exact reference to original intervals
		expect(updatedStore.workoutCreator.currentIntervals).toBe(
			store.workoutCreator.currentIntervals
		);
		expect(updatedStore.workoutCreator.history.length).toEqual(1);
		expect(updatedStore.workoutCreator.currentHistoryPosition).toEqual(0);
	});

	it('doesnt change intervals if set to same with extra properties', () => {
		const store = getInitialState();
		const intervalsWithExtra = store.workoutCreator.currentIntervals.map((i) => ({
			...i,
			extra: 100,
		}));

		const updatedStore = rootReducer(store, actions.setIntervals(intervalsWithExtra));

		// Confirm exact reference to original intervals
		expect(updatedStore.workoutCreator.currentIntervals).toBe(
			store.workoutCreator.currentIntervals
		);
		expect(updatedStore.workoutCreator.history.length).toEqual(1);
		expect(updatedStore.workoutCreator.currentHistoryPosition).toEqual(0);
	});

	it('undo redo', () => {
		const initialStore = getInitialState();

		const afterInitialUpdate = rootReducer(initialStore, actions.setIntervals([]));
		expect(afterInitialUpdate.workoutCreator.currentIntervals).toEqual([]);

		const afterUndo = rootReducer(afterInitialUpdate, actions.undo());
		expect(afterUndo.workoutCreator.currentIntervals).toEqual(
			initialStore.workoutCreator.currentIntervals
		);

		const afterRedo = rootReducer(afterUndo, actions.redo());
		expect(afterRedo.workoutCreator.currentIntervals).toEqual(
			afterInitialUpdate.workoutCreator.currentIntervals
		);
	});
});
