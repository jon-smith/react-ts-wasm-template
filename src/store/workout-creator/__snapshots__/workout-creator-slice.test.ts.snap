// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workout Creator Store initial state should match snapshot 1`] = `
Object {
  "currentHistoryPosition": 0,
  "currentIntervals": Array [
    Object {
      "durationSeconds": 60,
      "intensityPercent": 0.3,
    },
    Object {
      "durationSeconds": 60,
      "intensityPercent": 0.4,
    },
    Object {
      "durationSeconds": 60,
      "intensityPercent": 0.5,
    },
    Object {
      "durationSeconds": 300,
      "intensityPercent": 0.6,
    },
    Object {
      "durationSeconds": 30,
      "intensityPercent": 1.3,
    },
    Object {
      "durationSeconds": 15,
      "intensityPercent": 0.6,
    },
    Object {
      "durationSeconds": 30,
      "intensityPercent": 1.3,
    },
    Object {
      "durationSeconds": 15,
      "intensityPercent": 0.6,
    },
    Object {
      "durationSeconds": 30,
      "intensityPercent": 1.3,
    },
    Object {
      "durationSeconds": 15,
      "intensityPercent": 0.6,
    },
    Object {
      "durationSeconds": 30,
      "intensityPercent": 1.3,
    },
    Object {
      "durationSeconds": 15,
      "intensityPercent": 0.6,
    },
    Object {
      "durationSeconds": 30,
      "intensityPercent": 1.3,
    },
    Object {
      "durationSeconds": 15,
      "intensityPercent": 0.6,
    },
    Object {
      "durationSeconds": 30,
      "intensityPercent": 1.3,
    },
    Object {
      "durationSeconds": 15,
      "intensityPercent": 0.6,
    },
    Object {
      "durationSeconds": 30,
      "intensityPercent": 1.3,
    },
    Object {
      "durationSeconds": 15,
      "intensityPercent": 0.6,
    },
    Object {
      "durationSeconds": 30,
      "intensityPercent": 1.3,
    },
    Object {
      "durationSeconds": 15,
      "intensityPercent": 0.6,
    },
    Object {
      "durationSeconds": 30,
      "intensityPercent": 1.3,
    },
    Object {
      "durationSeconds": 15,
      "intensityPercent": 0.6,
    },
    Object {
      "durationSeconds": 30,
      "intensityPercent": 1.3,
    },
    Object {
      "durationSeconds": 15,
      "intensityPercent": 0.6,
    },
    Object {
      "durationSeconds": 30,
      "intensityPercent": 1.3,
    },
    Object {
      "durationSeconds": 15,
      "intensityPercent": 0.6,
    },
    Object {
      "durationSeconds": 30,
      "intensityPercent": 1.3,
    },
    Object {
      "durationSeconds": 15,
      "intensityPercent": 0.6,
    },
    Object {
      "durationSeconds": 30,
      "intensityPercent": 1.3,
    },
    Object {
      "durationSeconds": 15,
      "intensityPercent": 0.6,
    },
    Object {
      "durationSeconds": 300,
      "intensityPercent": 0.6,
    },
  ],
  "ftp": 200,
  "history": Array [
    Array [
      Object {
        "durationSeconds": 60,
        "intensityPercent": 0.3,
      },
      Object {
        "durationSeconds": 60,
        "intensityPercent": 0.4,
      },
      Object {
        "durationSeconds": 60,
        "intensityPercent": 0.5,
      },
      Object {
        "durationSeconds": 300,
        "intensityPercent": 0.6,
      },
      Object {
        "durationSeconds": 30,
        "intensityPercent": 1.3,
      },
      Object {
        "durationSeconds": 15,
        "intensityPercent": 0.6,
      },
      Object {
        "durationSeconds": 30,
        "intensityPercent": 1.3,
      },
      Object {
        "durationSeconds": 15,
        "intensityPercent": 0.6,
      },
      Object {
        "durationSeconds": 30,
        "intensityPercent": 1.3,
      },
      Object {
        "durationSeconds": 15,
        "intensityPercent": 0.6,
      },
      Object {
        "durationSeconds": 30,
        "intensityPercent": 1.3,
      },
      Object {
        "durationSeconds": 15,
        "intensityPercent": 0.6,
      },
      Object {
        "durationSeconds": 30,
        "intensityPercent": 1.3,
      },
      Object {
        "durationSeconds": 15,
        "intensityPercent": 0.6,
      },
      Object {
        "durationSeconds": 30,
        "intensityPercent": 1.3,
      },
      Object {
        "durationSeconds": 15,
        "intensityPercent": 0.6,
      },
      Object {
        "durationSeconds": 30,
        "intensityPercent": 1.3,
      },
      Object {
        "durationSeconds": 15,
        "intensityPercent": 0.6,
      },
      Object {
        "durationSeconds": 30,
        "intensityPercent": 1.3,
      },
      Object {
        "durationSeconds": 15,
        "intensityPercent": 0.6,
      },
      Object {
        "durationSeconds": 30,
        "intensityPercent": 1.3,
      },
      Object {
        "durationSeconds": 15,
        "intensityPercent": 0.6,
      },
      Object {
        "durationSeconds": 30,
        "intensityPercent": 1.3,
      },
      Object {
        "durationSeconds": 15,
        "intensityPercent": 0.6,
      },
      Object {
        "durationSeconds": 30,
        "intensityPercent": 1.3,
      },
      Object {
        "durationSeconds": 15,
        "intensityPercent": 0.6,
      },
      Object {
        "durationSeconds": 30,
        "intensityPercent": 1.3,
      },
      Object {
        "durationSeconds": 15,
        "intensityPercent": 0.6,
      },
      Object {
        "durationSeconds": 30,
        "intensityPercent": 1.3,
      },
      Object {
        "durationSeconds": 15,
        "intensityPercent": 0.6,
      },
      Object {
        "durationSeconds": 300,
        "intensityPercent": 0.6,
      },
    ],
  ],
  "newInterval": Object {
    "durationSeconds": 0,
    "intensityPercent": 1,
  },
  "selectedIndex": null,
}
`;
